File: python/koans/24_river_crossing/test_river.py
import random
import threading
import time

from koan_utils import (
    EventLog,
    ThreadRunner,
    eventually,
    jitter,
)

from river import Boat


class BoatHooks:
    """Records boards and rows; dawdles inside board() to widen windows."""

    def __init__(self):
        self.log = EventLog()

    def board(self, kind):
        jitter(2.0)
        self.log.record(f"board:{kind}")

    def row_boat(self, kind):
        self.log.record("row")

    def boards(self):
        return [e.split(":", 1)[1] for e in self.log.events() if e.startswith("board:")]

    def rows(self):
        return self.log.count("row")


def assert_boatloads_legal(boards):
    assert len(boards) % 4 == 0, (
        f"board count {len(boards)} is not a multiple of 4: {boards}"
    )
    for i in range(0, len(boards), 4):
        block = boards[i : i + 4]
        h, s = block.count("hacker"), block.count("serf")
        assert (h, s) in ((4, 0), (0, 4), (2, 2)), (
            f"boatload #{i // 4} is illegal: {block} (full sequence: {boards})"
        )


def assert_full_pattern(events):
    """Each boatload: exactly 4 boards, then exactly one row, no overlap."""
    pending = 0
    for e in events:
        if e.startswith("board:"):
            pending += 1
            assert pending <= 4, (
                f"a fifth passenger boarded before the boat sailed: {events}"
            )
        elif e == "row":
            assert pending == 4, (
                f"row_boat fired before all four passengers boarded: {events}"
            )
            pending = 0
    assert pending == 0, f"a boatload boarded but never sailed: {events}"


def run_mix(n_hackers, n_serfs, join_timeout=15):
    hooks = BoatHooks()
    boat = Boat(hooks)
    runner = ThreadRunner()
    kinds = ["hacker"] * n_hackers + ["serf"] * n_serfs
    random.shuffle(kinds)
    for i, kind in enumerate(kinds):
        def body(kind=kind):
            jitter()
            (boat.hacker_arrives if kind == "hacker" else boat.serf_arrives)()

        runner.spawn(body, name=f"{kind}-{i}")
    runner.join_all(timeout=join_timeout)
    return hooks


def test_three_and_one_never_sails():
    hooks = BoatHooks()
    boat = Boat(hooks)
    done = EventLog()

    def rider(kind):
        try:
            (boat.hacker_arrives if kind == "hacker" else boat.serf_arrives)()
            done.record(kind)
        except BaseException as exc:  # surfaced by the asserts below
            done.record(f"error: {exc!r}")

    for _ in range(3):
        threading.Thread(target=rider, args=("hacker",), daemon=True).start()
    threading.Thread(target=rider, args=("serf",), daemon=True).start()
    time.sleep(0.4)  # 3 hackers + 1 serf: give an illegal crew every chance
    assert hooks.log.events() == [], (
        f"3 hackers + 1 serf must not board anything, saw {hooks.log.events()}"
    )
    assert done.events() == [], (
        f"nobody may cross in an illegal combination: {done.events()}"
    )
    # A fourth hacker makes an all-hacker crew possible; the serf stays.
    threading.Thread(target=rider, args=("hacker",), daemon=True).start()
    eventually(
        lambda: done.count("hacker") == 4,
        timeout=5,
        msg="four hackers should sail together once the fourth arrives",
    )
    assert hooks.boards() == ["hacker"] * 4, (
        f"expected exactly the 4 hackers to board, saw {hooks.boards()}"
    )
    assert hooks.rows() == 1
    time.sleep(0.3)
    assert done.count("serf") == 0, "the lone serf must keep waiting ashore"


def test_pairs_combination_sails():
    hooks = run_mix(2, 2, join_timeout=5)
    assert sorted(hooks.boards()) == ["hacker", "hacker", "serf", "serf"]
    assert hooks.rows() == 1


def test_four_of_a_kind_sails():
    hooks = run_mix(0, 4, join_timeout=5)
    assert hooks.boards() == ["serf"] * 4
    assert hooks.rows() == 1


def test_rowing_after_all_board():
    for n_h, n_s in ((2, 2), (4, 0), (0, 4)):
        hooks = run_mix(n_h, n_s, join_timeout=5)
        events = hooks.log.events()
        assert hooks.rows() == 1, f"exactly one rower per boatload: {events}"
        row_index = events.index("row")
        board_indexes = [i for i, e in enumerate(events) if e.startswith("board:")]
        assert len(board_indexes) == 4
        assert all(i < row_index for i in board_indexes), (
            f"row_boat must come after all four boards: {events}"
        )


def test_boatloads_do_not_interleave():
    hooks = run_mix(8, 8)
    assert_boatloads_legal(hooks.boards())
    assert hooks.rows() == 4
    assert_full_pattern(hooks.log.events())


def test_stress():
    hooks = run_mix(16, 16, join_timeout=20)
    assert len(hooks.boards()) == 32
    assert_boatloads_legal(hooks.boards())
    assert hooks.rows() == 8
    assert_full_pattern(hooks.log.events())

File: cpp/koans/24_river_crossing/test_river.cpp
#include "koan_test.hpp"
#include "river.hpp"

#include <algorithm>
#include <memory>
#include <random>
#include <string>
#include <thread>
#include <vector>

using namespace koans;

namespace {

struct Bench {
    std::shared_ptr<EventLog> log = std::make_shared<EventLog>();
    std::shared_ptr<Boat> boat;

    Bench() {
        auto l = log;
        boat = std::make_shared<Boat>(BoatHooks{
            [l](const std::string& kind) {
                jitter(2);
                l->record("board:" + kind);
            },
            [l](const std::string&) { l->record("row"); }});
    }

    std::vector<std::string> boards() const {
        std::vector<std::string> out;
        for (const auto& e : log->events())
            if (e.rfind("board:", 0) == 0) out.push_back(e.substr(6));
        return out;
    }

    std::size_t rows() const { return log->count("row"); }
};

void assert_boatloads_legal(const std::vector<std::string>& boards) {
    KOAN_ASSERT_MSG(boards.size() % 4 == 0,
                    "board count " + std::to_string(boards.size()) +
                        " is not a multiple of 4");
    for (std::size_t i = 0; i < boards.size(); i += 4) {
        int h = 0, s = 0;
        for (std::size_t j = i; j < i + 4; ++j) {
            if (boards[j] == "hacker") ++h;
            if (boards[j] == "serf") ++s;
        }
        bool legal = (h == 4 && s == 0) || (h == 0 && s == 4) || (h == 2 && s == 2);
        KOAN_ASSERT_MSG(legal, "boatload #" + std::to_string(i / 4) +
                                   " is illegal: " + std::to_string(h) +
                                   " hackers + " + std::to_string(s) + " serfs");
    }
}

// Each boatload: exactly 4 boards, then exactly one row, no overlap.
void assert_full_pattern(const std::vector<std::string>& events) {
    int pending = 0;
    for (const auto& e : events) {
        if (e.rfind("board:", 0) == 0) {
            ++pending;
            KOAN_ASSERT_MSG(pending <= 4,
                            "a fifth passenger boarded before the boat sailed");
        } else if (e == "row") {
            KOAN_ASSERT_MSG(pending == 4,
                            "row_boat fired before all four passengers boarded");
            pending = 0;
        }
    }
    KOAN_ASSERT_MSG(pending == 0, "a boatload boarded but never sailed");
}

Bench run_mix(int n_hackers, int n_serfs,
              std::chrono::milliseconds join_timeout = 15000ms) {
    Bench bench;
    ThreadRunner runner;
    std::vector<char> kinds(static_cast<std::size_t>(n_hackers), 'h');
    kinds.insert(kinds.end(), static_cast<std::size_t>(n_serfs), 's');
    std::mt19937 rng{std::random_device{}()};
    std::shuffle(kinds.begin(), kinds.end(), rng);
    auto boat = bench.boat;
    for (char kind : kinds) {
        runner.spawn([boat, kind] {
            jitter();
            if (kind == 'h')
                boat->hacker_arrives();
            else
                boat->serf_arrives();
        });
    }
    runner.join_all(join_timeout);
    return bench;
}

}  // namespace

KOAN_TEST(three_and_one_never_sails) {
    Bench bench;
    auto boat = bench.boat;
    auto log = bench.log;
    auto done = std::make_shared<EventLog>();

    auto spawn_rider = [boat, done](const std::string& kind) {
        // Detached: the lone serf stays parked past the end of the test,
        // so everything it touches is shared_ptr-owned.
        std::thread([boat, done, kind] {
            try {
                if (kind == "hacker")
                    boat->hacker_arrives();
                else
                    boat->serf_arrives();
                done->record(kind);
            } catch (const std::exception& e) {
                done->record(std::string("error: ") + e.what());
            }
        }).detach();
    };

    for (int i = 0; i < 3; ++i) spawn_rider("hacker");
    spawn_rider("serf");
    std::this_thread::sleep_for(400ms);  // give an illegal crew every chance
    KOAN_ASSERT_MSG(log->events().empty(),
                    "3 hackers + 1 serf must not board anything, saw " +
                        log->joined());
    KOAN_ASSERT_MSG(done->events().empty(),
                    "nobody may cross in an illegal combination: " +
                        done->joined());
    // A fourth hacker makes an all-hacker crew possible; the serf stays.
    spawn_rider("hacker");
    eventually([done] { return done->count("hacker") == 4; }, 5000ms,
               "four hackers should sail together once the fourth arrives");
    auto boards = bench.boards();
    KOAN_ASSERT_MSG(boards == std::vector<std::string>(4, "hacker"),
                    "expected exactly the 4 hackers to board");
    KOAN_ASSERT_EQ(bench.rows(), std::size_t{1});
    std::this_thread::sleep_for(300ms);
    KOAN_ASSERT_MSG(done->count("serf") == 0,
                    "the lone serf must keep waiting ashore");
}

KOAN_TEST(pairs_combination_sails) {
    auto bench = run_mix(2, 2, 5000ms);
    auto boards = bench.boards();
    std::sort(boards.begin(), boards.end());
    KOAN_ASSERT(boards ==
                (std::vector<std::string>{"hacker", "hacker", "serf", "serf"}));
    KOAN_ASSERT_EQ(bench.rows(), std::size_t{1});
}

KOAN_TEST(four_of_a_kind_sails) {
    auto bench = run_mix(0, 4, 5000ms);
    KOAN_ASSERT(bench.boards() == std::vector<std::string>(4, "serf"));
    KOAN_ASSERT_EQ(bench.rows(), std::size_t{1});
}

KOAN_TEST(rowing_after_all_board) {
    struct Mix { int h, s; };
    for (Mix mix : {Mix{2, 2}, Mix{4, 0}, Mix{0, 4}}) {
        auto bench = run_mix(mix.h, mix.s, 5000ms);
        auto events = bench.log->events();
        KOAN_ASSERT_MSG(bench.rows() == 1, "exactly one rower per boatload");
        long row_index = -1;
        std::vector<long> board_indexes;
        for (std::size_t i = 0; i < events.size(); ++i) {
            if (events[i] == "row") row_index = static_cast<long>(i);
            if (events[i].rfind("board:", 0) == 0)
                board_indexes.push_back(static_cast<long>(i));
        }
        KOAN_ASSERT_EQ(board_indexes.size(), std::size_t{4});
        for (long b : board_indexes)
            KOAN_ASSERT_MSG(b < row_index,
                            "row_boat must come after all four boards: " +
                                bench.log->joined());
    }
}

KOAN_TEST(boatloads_do_not_interleave) {
    auto bench = run_mix(8, 8);
    assert_boatloads_legal(bench.boards());
    KOAN_ASSERT_EQ(bench.rows(), std::size_t{4});
    assert_full_pattern(bench.log->events());
}

KOAN_TEST(stress) {
    auto bench = run_mix(16, 16, 20000ms);
    KOAN_ASSERT_EQ(bench.boards().size(), std::size_t{32});
    assert_boatloads_legal(bench.boards());
    KOAN_ASSERT_EQ(bench.rows(), std::size_t{8});
    assert_full_pattern(bench.log->events());
}
